File: 92.py
#!/usr/bin/env python

"""This script solves the Project Euler problem "Square digit chains". The
problem is: How many starting numbers below ten million will arrive at 89?
"""


def main():
    """Square digit chains"""

    # Constants
    MAX = 10000000

    count = 0

    cache = {
        '1': 1,
        '89': 89,
    }

    for num in range(2, MAX):
        chain = []
        final = None
        n = ''.join(sorted(digit for digit in str(num)))
        while (True):
            if n in cache:
                final = cache[n]
                break
            chain.append(n)
            n = str(sum(int(digit) * int(digit) for digit in str(n)))
        if final == 89:
            count += 1
        for n in chain:
            cache[n] = final

    print(count)

if __name__ == '__main__':
    main()

File: 92.js
#!/usr/bin/env node

// Square digit chains

var MAX = 10000000;

var count = 0;

cache = {
    '1': 1,
    '89': 89,
};

for (var num = 2; num < MAX; num++) {
    var chain = [];
    var final;
    var n = num.toString().split('').sort().join('');
    while (true) {
        if (n in cache) {
            final = cache[n];
            break;
        }
        chain.push(n);
        var sum = 0;
        digits = n.toString().split('');
        for (var i = 0; i < digits.length; i++) {
            sum += parseInt(digits[i]) * parseInt(digits[i]);
        }
        n = sum;
    }
    if (final == 89) {
        count++;
    }
    for (var i = 0; i < chain.length; i++) {
        cache[n] = final;
    }
}

console.log(count);
